ext/liquid_c/block.c

```diff
@@ -1,59 +1,142 @@
 {
-    return rb_isalnum(c) || c == '_';
+    block_body_t *body = ptr;
+    rb_gc_mark(body->source);
+
+    size_t *const_ptr = (size_t *)body->constants.data;
+    const uint8_t *ip = body->instructions.data;
+    // Don't rely on a terminating OP_LEAVE instruction
+    // since this could be called in the middle of parsing
+    const uint8_t *end_ip = ip + body->instructions.size;
+    while (ip < end_ip) {
+        switch (*ip++) {
+            case OP_LEAVE:
+                break;
+            case OP_WRITE_RAW:
+                const_ptr += 2;
+                break;
+            case OP_WRITE_NODE:
+                rb_gc_mark(*const_ptr++);
+                break;
+            default:
+                rb_bug("invalid opcode: %u", ip[-1]);
+        }
+    }
 }
 
-static VALUE yield_end_tag(VALUE tag_name, VALUE tag_markup, VALUE parse_context, bool for_liquid_tag)
+static void block_body_free(void *ptr)
 {
-    if (!for_liquid_tag) {
-        return rb_yield_values(2, tag_name, tag_markup);
-    }
+    block_body_t *body = ptr;
+    c_buffer_free(&body->instructions);
+    c_buffer_free(&body->constants);
+    xfree(body);
+}
 
-    return rb_funcall(cLiquidBlockBody, intern_unknown_tag_in_liquid_tag, 2, tag_name, parse_context);
+static size_t block_body_memsize(const void *ptr)
+{
+    const block_body_t *body = ptr;
+    if (!ptr) return 0;
+    return sizeof(block_body_t) + body->instructions.capacity + body->constants.capacity;
 }
 
-static VALUE internal_block_parse(VALUE self, VALUE tokens, VALUE parse_context, bool for_liquid_tag)
+const rb_data_type_t block_body_data_type = {
+    "liquid_block_body",
+    { block_body_mark, block_body_free, block_body_memsize, },
+    NULL, NULL, RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+#define BlockBody_Get_Struct(obj, sval) TypedData_Get_Struct(obj, block_body_t, &block_body_data_type, sval)
+
+static inline void block_body_write_opcode(block_body_t *body, enum opcode op)
 {
-    tokenizer_t *tokenizer;
-    Tokenizer_Get_Struct(tokens, tokenizer);
+    c_buffer_write(&body->instructions, &op, 1);
+}
 
+static inline void block_body_add_leave(block_body_t *body)
+{
+    block_body_write_opcode(body, OP_LEAVE);
+}
+
+static void block_body_add_write_raw(block_body_t *body, const char *string, size_t size)
+{
+    block_body_write_opcode(body, OP_WRITE_RAW);
+    size_t slice[2] = { (size_t)string, size };
+    c_buffer_write(&body->constants, &slice, sizeof(slice));
+}
+
+static void block_body_add_write_node(block_body_t *body, VALUE node)
+{
+    block_body_write_opcode(body, OP_WRITE_NODE);
+    c_buffer_write(&body->constants, &node, sizeof(VALUE));
+}
+
+static VALUE block_body_allocate(VALUE klass)
+{
+    block_body_t *body;
+
+    VALUE obj = TypedData_Make_Struct(klass, block_body_t, &block_body_data_type, body);
+    body->instructions = c_buffer_allocate(8);
+    block_body_add_leave(body);
+    body->constants = c_buffer_allocate(8 * sizeof(VALUE));
```

Comment:
It was chosen fairly arbitrarily based on the current situation where we could end up with a lot of small block bodies for things like `if` tags.  I tried to keep it as a multiple of 2, since that seemed like it would make it more likely that the memory allocator will re-use buffer allocations for new or expanding buffers and because it could use whole pages for larger memory allocations.  However, I haven't tested any of these assumptions yet.

We will probably want to revisit the default Liquid::Document's block body as we start compiling more functionality into VM instructions, since then we will be able to compile builtin block tags into their parent block body.